cupy/cuda/memory_hook.py

```diff
@@ -0,0 +1,96 @@
+import collections
+import threading
+
+thread_local = threading.local()
+
+
+def get_memory_hooks():
+    if not hasattr(thread_local, 'memory_hooks'):
+        thread_local.memory_hooks = collections.OrderedDict()
+    return thread_local.memory_hooks
+
+
+class MemoryHook(object):
+    """Base class of hooks for Memory allocations.
+
+    :class:`~cupy.cuda.MemoryHook` is an callback object
+    that is registered to :class:`~cupy.cuda.SingleDeviceMemoryPool`.
```

Comment:
I am afraid I do not have a concrete use case in which users want to append different memory hooks for different devices, although I made the comment like that. It would be helpful if you shared with me an example if you have in mind.